### PR TITLE
[Merged by Bors] - feat: rewrite the windows line ending linter in Lean

### PR DESCRIPTION
**Pull request by grunweg**
This linter is different from the other text-based linters, as it does not operate on the output of `IO.FS.lines`: since Lean 4.11, that function treats windows and unix line endings the same, hence its result is not useful here. Instead, we read the file contents and use `crlfToLf` to detect windows endings.

We consciously do not convert this to a syntax-based linter: if a file has bad line endings, it is plausible that many line endings are unexpected, and the best fix is automatic (not manual, as a syntax linter would prompt one to).

This is the first text-based linter which includes an automatic fix: hence, we also add the corresponding infrastructure to perform such fixes in general.

### CODE REVIEW

File: Mathlib/Tactic/Linter/TextBased.lean [hunk 1]
```diff
@@ -213,8 +219,9 @@ def formatErrors (errors : Array ErrorContext) (style : ErrorFormat) : IO Unit :
     IO.println (outputMessage e style)
 
 /-- Core logic of a text based linter: given a collection of lines,
-return an array of all style errors with line numbers. -/
-abbrev TextbasedLinter := Array String → Array (StyleError × ℕ)
+return an array of all style errors with line numbers and, optionally, an array of lines
+changed to fix the linter automatically. (This is not always possible, but sometimes it is.) -/
```
Comment (semorrison) [inline Mathlib/Tactic/Linter/TextBased.lean:223]: The new optional array being returned... is this all the lines of the file? Just lines that were changed? If so, which? Hopefully we just need a better doc-string here.

File: scripts/lint-style.lean [hunk 2]
```diff
@@ -22,6 +22,10 @@ def lintStyleCli (args : Cli.Parsed) : IO UInt32 := do
     | (true, _) => OutputSetting.update
     | (false, true) => OutputSetting.print ErrorFormat.github
     | (false, false) => OutputSetting.print ErrorFormat.humanReadable
+  let fix := args.hasFlag "fix"
```
Comment (semorrison) [inline scripts/lint-style.lean:25]: Use this variable below rather than looking it up again.

File: Mathlib/Tactic/Linter/TextBased.lean (hunk 1 above)
Comment (semorrison) [reply Mathlib/Tactic/Linter/TextBased.lean:223]: If it's just the whole file, would it made sense to intercalate all the lines with `\n` and just return a single string?

Comment (grunweg) [reply Mathlib/Tactic/Linter/TextBased.lean]: It's all the lines of the file. Is the new docstring clear enough?

It seems tempting, but the answer is no: if there is more than one auto-fixable text-based linter (and there will be in the future), one would like the auto-fixes to compound --- fixing first one error, then the other etc. In particular, later linters need to work with the result of auto-fixing, thus need an array of lines to work with. Intercalating right away would not help, as one would only need to split the string again.

**Comment by grunweg** [comment]
Fixed the merge conflict.

File: Mathlib/Tactic/Linter/TextBased.lean [hunk 3]
```diff
@@ -213,8 +219,11 @@ def formatErrors (errors : Array ErrorContext) (style : ErrorFormat) : IO Unit :
     IO.println (outputMessage e style)
 
 /-- Core logic of a text based linter: given a collection of lines,
-return an array of all style errors with line numbers. -/
-abbrev TextbasedLinter := Array String → Array (StyleError × ℕ)
+return an array of all style errors with line numbers and, optionally,
+the collection of lines changed to fix the linter errors automatically.
+(Such automatic fixes are only possible for some kinds of `StyleError`s.)
```
Comment (semorrison) [inline Mathlib/Tactic/Linter/TextBased.lean:224]: This doc-string is still ambiguous about which lines are returned.

**Comment by semorrison** [comment]
bors d+

File: Mathlib/Tactic/Linter/TextBased.lean (hunk 3 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/TextBased.lean]: Indeed... clarified the docstring further.

**Comment by grunweg** [comment]
Thanks for the review!
bors merge

**Pull request closed without merging**